### DISCUSSION TITLE
Improve readability of Meltano's console output

### DISCUSSION THREAD

**Discussion by aaronsteers** [meltano/meltano#7016] [Ideas]
Starting this discussion to scope out work that would improve the 'default' user experience when running Meltano.

Related: 

- https://github.com/meltano/meltano/discussions/6642#discussioncomment-3448447

A few changes I'll propose for a start, and folks can chime in with comments below for other ideas/proposals.

## Enable file logging by default

- https://github.com/meltano/meltano/issues/7081

<details>
<summary>Details</summary>

By default, we should _enable_ structlog log files to always be sent to `.meltano/logs` (aka `$MELTANO_SYS_DIR_ROOT/logs`)

Today:
- A full log appears to go under `.meltano/logs/elt/{state-id}/{uuid}/elt.log`. This is bad because the state ID may not even apply, and the job may not even be an `elt` run.
- The log appears to only contain log messages from plugins subprocesses - and is importantly missing log data from Meltano itself.

Minimal solve:

1. Everything that today would be printed to the console would all go to the same log file, whether from Meltano itself or from a Plugin.
1. The log file would be named something like `.meltano/logs/meltano/yyyy-mm-dd/hh-mm-ss-{uuid}.log`, with timestamps keyed off of the starting timestamp. It should not contain the state ID as it does today.

Additional solve for plugin-specific or command-specific logs:

1. Meltano could create additional log files, one for each plugin subprocesses. If something fails within that plugin, we print the path to the log _just_ for plugin that failed.
1. Following a convention like above, a plugin-specific or command-specific log file would be named something like `.meltano/logs/{plugin}/yyyy-mm-dd/hh-mm-ss-{uuid}.log`.

</details>

## Point users to log files when helpful

Also tracked in:

- #7081 

When execution fails:

- If the failure is in a plugin subprocess: print the path to the plugin-specific log.
- Else, if the failure is a `MeltanoError`, do not print the path.
  - Excepting Plugin subprocess failure, log path is not needed for handled errors like like `PluginNotFound: 'tap-gitlaab' is not regognized. Please check the spelling and try again.`
- Else, if the failure is in Meltano itself (an unhandled Exception, for instance): print the path to the 'full' execution log.

## Clean up Meltano's own output

1. Disable `structlog` formatting in the console by default - unless we confirm a non-tty interface or unless the user has a logging.yaml that gives different instructions.
   - Since we're directing users to the structlog log file if/when a failure occurs in Meltano or in its subprocesses, we don't need to also echo all of that information to the screen.

## Revert to old EL output in the console

1. For EL operations, or any time where we have multiple plugins running in parallel: echo the stderr directly using the plugin name as a prefix.
1. As Meltano did previously, we can color the plugin names prefix text differently per plugin (if `NO_COLOR` is not set), so that plugins' respective output is more easily distinguished when all are sending output at the same time.
   - E.g. similar to what we previously had:
    ```
    tap-snowflake | Starting sync
    tap-snowflake | Synced 100 rows
    ...
    my-mapper     | Transformed 100 rows
    ...
    target-csv    | Received 100 rows
    ...

## Clean up output for general plugin commands

1. At the start of invoking any plugin, we tell the user that what follows is from the plugin and not from us, something like: `---------- Starting my-cool-plugin... ----------`.
1. We echo to the console exactly what we're getting from the plugin, with no adornments or extra log artifacts.
   - This keeps log output for something like dbt to be _exactly_ as readable as it would be if the user invoked dbt directly.
   - Because we prefixed the block with the name of the plugin and/or command we're running, and because we don't have multiple plugins sending output at the same time, we don't need any additional text saying which plugin generated the output.
   - If the user wants timestamps of individual steps, they can refer to the structlog log file.
1. We'd end the output stream with something that says the plugin completed and how long it took: `---------- Plugin my-cool-plugin completed. Total runtime 4.3 minutes. --------`. The pre-print and post-print test replaces the more verbose structlog wrapper, while still giving user necessary information on which plugin or process is generating the output and how long (once it is complete) the process itself took.

## Improve exception handling and error handling

1. We should ensure that which are subclasses of `MeltanoError` do not print a stack trace. Instead those should include helpful output: what happened, how to fix it, how to get help. (E.g. "InvalidPluginNameError. Meltano could not find a plugin named 'tap-gitlaab'. Please check the spelling and try again. If you need help, please visit https://melta.no/...") We can optionally print a path to the execution log, but no stack trace or other log info should need to be printed to the console.
1. We should add catch/except logic for at least the 'top 10' exception instances by frequency, ensuring that each raises a proper `MeltanoError` instead of the triggering `Exception`.

## "Smart" EL Status Indicators

In a later phase we can go back and change EL status indicators altogether, introduce new libraries, and try to achieve a more 'joyful' experience overall.

1. We would no longer echo tap and target logs directly, at least not during the sync operations. Likely for `--discover` operations, we'd run those as described below for 'other plugins and commands'.
3. Instead of printing tap/target logs as they are emitted, we print one of these:
   1. An simple animated `...` or a more advanced 'progress bar' animation. Since we don't know the total duration, we'd just be indicating 'progress of some kind' and not 'percentage of work'. 
   4. Perhaps better: a rolling record-count tally, either overall or per stream, using the data from the metrics log output on STDERR (doesn't require logging or interjecting in the main STDOUT stream.

**Comment by aaronsteers** [comment]
@tayloramurphy, @pnadolny13, @sbalnojan - I've taken several rounds of iteration on the above and would love feedback/discussion on these points.

The above is very detailed so here is the per-line-item summary:

1. **Enable file logging by default.** This helps users debug, but it also removes the need for console-level output to include everything under the sun. 
2. **Point users to log files when helpful.** If there's a failure, we can point users to the log file. The prevents us from relying too much on console output.
3. **Clean up Meltano's own output.** Once we have log files capturing detailed execution information, we can disable `structlog` in the console, unless the user chooses to configure it that way in `logging.yaml`.
4. **Revert to old EL output in the console.** Our old EL output was actually easier to read. We should probably revert to that as the default.
5. **Clean up output for general plugin commands.** We should always tell the user when a plugin is beginning to execute and when it has finished. In between, we should just echo the plugin's output and not inject a bunch more noise that makes native output harder to utilize.
6. **Improve exception handling and error handling.** "Errors" are things the user probably needs to fix or deal with, like a mispelling plugin name or failed tap auth, or calling a command that doesn't exist. "Exceptions" are places where our code doesn't know what to do and so it crashes and prints out a stack trace of everything that went wrong. We should try to make sure common issues are handled and messaged properly as Errors and not as Exceptions.
7. **"Smart" EL Status Indicators.** Rather than print 10 million lines of EL log data while things are syncing, we could just print a `Syncing...` message, perhaps with an infinite ellipse or status bar to prove the program isn't 'stuck'. We could even print the path to the log file(s) at the very beginning of the sync operation, so users could `tail` it if they really want to see the detail.

What do you think of these? Should we break into 7+ issues and try to prioritize?

**Reply by tayloramurphy** [reply]
@aaronsteers this is well thought out and very clear. +1 to opening new issues and trying to prioritize. 

> Revert to old EL output in the console. Our old EL output was actually easier to read. We should probably revert to that as the default.

Let's reframe that as just taking another iteration on the console output once we have the log files. I don't want us to just revert, but we should redesign it given the context of everything we know since we brought on the structured logs.

**Reply by pnadolny13** [reply]
I agree 100%, very well thought out and I think these improvements would make a big difference.

**Comment by sbalnojan** [comment]
@aaronsteers Love the structure. 

What I (as usr) would  love from an error message is pretty simple, I would like to get an idea on what I have to fix.

That makes me think we should start by getting some data on the most common errors and see whether we can abstract from there on and map them onto your structure. @pnadolny13 I think we actually do have that right? I do remember some of the top problems being problems with Postgres right?

**Reply by pnadolny13** [reply]
@sbalnojan I created https://github.com/meltano/squared/issues/492 to dig into the exception telemetry data that we have to see what I can find. I've never looked into that data much and a lot of its hashed so its hard to glean much from it before parsing and unhashing more.

**Comment by sbalnojan** [comment]
@aaronsteers I got a question, my personal like most major problem with Meltanos console output is simply, that is is often "wrong". 

I just stumbled over an example, so maybe you can help me understand whether "wrong output/error messages" are already covered above:


If I have a target-postgres already installed, and then run a meltano add target-postgres --variant, I get this error:

`# meltano add loader target-postgres --variant meltanolabs
Loader 'target-postgres' already exists in your Meltano project
Need help fixing this problem? Visit http://melta.no/ for troubleshooting steps, or to
join our friendly Slack community.

Loader 'target-postgres' variant 'meltanolabs' is not known to Meltano. Variants: transferwise (default)`

And that did make me check the hub to see whether something is wrong. I started to wonder why I cannot have two variants, and so on. Until I deleted my current variant and realized the output simply was wrong. 

So I guess I'm just curious in what category "wrong error messages" falls above?

**Reply by aaronsteers** [reply]
@sbalnojan - The discussion scope here only covers failures which don't already have a Meltano Error being raised. Your example does have a Meltano Error raised but the message or inferred cause of failure is incorrect. It's even possible that our catching of exceptions and raising of proper error messages _contributes_ to the mismatched error message problem in the short run, because where we previously just let everything float up and print a huge stack trace, we're now prioritizing interpreted and properly messaged errors, which inevitably we'll get wrong sometimes.

The best solution then for the "wrong error message" problem is just to immediately log these as bugs. They'll almost always have a very small `weight`, especially when repro steps are available. We could also discuss prioritizing these small wins highly to get them completed in a quick turnaround.

Does this answer your question? And is this a good path forward?

**Comment by edgarrmondragon** [comment]
Things are significantly better now imo
